node/src/service/p2p/bootstrap.rs

```diff
@@ -61,46 +104,166 @@ impl BootstrapConfig {
 pub(crate) async fn bootstrap(
     mut swarm: Swarm<BootstrapBehaviour>,
     addr: Multiaddr,
+    bootstrap_addresses: Vec<Multiaddr>,
 ) -> Result<(), P2PError> {
     info!("Starting P2P bootstrap node at {addr}");
     swarm.listen_on(addr)?;
+    for addr in bootstrap_addresses {
+        info!("Dialing {addr}");
+        swarm.dial(addr)?;
+    }
+    swarm
+        .behaviour_mut()
+        .gossipsub
+        .subscribe(&IdentTopic::new(GOSSIP_TOPIC))?;
+    let mut registrations = HashMap::new();
     loop {
-        match swarm.select_next_some().await {
-            SwarmEvent::NewListenAddr { address, .. } => {
-                info!("Listening on {}", address);
-            }
-            SwarmEvent::Behaviour(BootstrapBehaviourEvent::Rendezvous(
-                rendezvous::server::Event::PeerRegistered { peer, registration },
-            )) => {
-                info!(
-                    "Peer {} registered for namespace '{}' for {} seconds",
-                    peer, registration.namespace, registration.ttl
-                );
+        tokio::select! {
+            event = swarm.select_next_some() => match event {
+                SwarmEvent::NewListenAddr { address, .. } => {
+                    info!("Listening on {}", address);
+                }
+                SwarmEvent::ConnectionEstablished { peer_id, .. } => {
+                    info!("Connected to {}", peer_id);
+                }
+                SwarmEvent::ConnectionClosed { peer_id, .. } => {
+                    info!("Disconnected from {}", peer_id);
+                }
+                SwarmEvent::Behaviour(BootstrapBehaviourEvent::Rendezvous(event)) => on_rendezvous_event(&mut swarm, event, &mut registrations),
+                SwarmEvent::Behaviour(BootstrapBehaviourEvent::Gossipsub(event)) => on_gossipsub_event(event, *swarm.local_peer_id(), &mut registrations),
+                SwarmEvent::Behaviour(BootstrapBehaviourEvent::RequestResponse(event)) => on_request_response_event(&mut swarm, event, &registrations),
+                other => debug!("Encountered event: {other:?}"),
```

Comment:
That's funny that we just can pass shared exclusive references to those things. 
It means that we're handling everything in the same task.
I guess for know we don't need to introduce the event loop pattern with spawning on messages and synchronization, because all of the message handling is so fast that it won't be blocking?

But then if we're blocking the thread on `RequestResponse`, then a client which is purposefully slow with its responses can DoS the other messages handling.

Reply:
Ok, i'm fine with that then.